### ISSUE TITLE
Bundle chromium renderer

### ISSUE THREAD

**Issue by Sytten** [tauri-apps/tauri#14963]
**Is your feature request related to a problem? Please describe.**
The performances of WebkitGTK are really bad and the implementation often breaks. It is also hard to test for app maintainers since nobody really uses it outside of some obscure default browser. It is also not on the MDN compatibility list for API.
For all those reasons, there should be an alternative  

**Describe the solution you'd like**
For better or worst, shipping a chromium renderer like electron does is the way all big companies ship their app today and the reason is simple: it works great. The experience is consistent for the user and it's not a pain to QA.
I think wry should support bundling a renderer.

**Describe alternatives you've considered**
I know that there is talk to integrate servo into it, but honestly it doesn't solve the compatibility and QA issue outlined above.
It's not a bad idea per say, so if it can be done in a modular way (feature flag `servo` and `chromium` with some traits on top) I think that would be great.

**Would you want to assign yourself to implement this feature?**
I am happy to contribute changes since I kinda know the codebase a bit by now, but it's a big change. I don't expect this issue to move the team on a dime, but I wanted to start the discussion.

**Comment by MaksymSharinDev** [comment]
This is the crate that must be pushed forward to make it possible I guess
https://github.com/wusyong/cef-rs

Or hoping for a very far in time linux support  for WebView2 
https://github.com/MicrosoftEdge/WebView2Feedback/issues/645

**Comment by Sytten** [comment]
I am circling back on this because it causes us so much pain on Linux that we are always discussing dropping Tauri because of it. I cannot overstate how bad webkitgtk performances are compared to browsers.

**Comment by FabianLars** [comment]
We're aware of the situation and completely stopped defending webkitgtk ourselves. Sadly it is still a huge amount of work that we currently don't seem to be able to stem... Nonetheless, we do not see a future with webkitgtk.

And i agree that using Tauri, or any similar framework using webkitgtk, is not a good idea for projects where Linux is a serious target, to a degree where i actively warn people about this (before the never ending webkitgtk-nvidia fiasco began i restricted this warning to performance intensive frontends but sadly we're past that now).

----

My idea was to work on the webkitgtk 6.0 port (basically webkitgtk with gtk4) for tauri v3 and then some point after that look into cef again (my qt webview experiments never seem to work out 😩) if servo needs more time, but i've since changed my mind and will try my best to start working on an alternative for webkitgtk after v2 is stable. Hopefully to a degree where we can completely drop it but that may be a bit too optimistic. I also hope that @simonhyll is still interested in working on this too then ❤️ 

----

In the meantime, we can only apologize for any inconveniences (or worse) this caused any of you.

**Comment by Sytten** [comment]
Thanks for the answer @FabianLars, you guys are fighting the good fight. Honestly if Microsoft had any vision of would be awesome to have a webview2 standard across OS and it could be even a user setting (like use Firefox to provide webviews in my apps). Sadly we are not there yet. 

Seems like the latest webkitgtk shipped with gtk 4 support so there might be some improvements going there. But before that ships in the OS it is going to be years. That is kind of my main grip with native webview design, you often encounter really really outdated webviews even on recent machines and most of the time the user can't update it easily.

CEF work I think is the way to go. Even better if the user can patch the chromium version without us. I know that for security oriented OS (like Parrot OS) the maintenance of electron apps is annoying because they cant easily patch the browser in case there is a critical CVE.

**Comment by FabianLars** [comment]
> Honestly if Microsoft had any vision of would be awesome to have a webview2 standard across OS

Yeah, we hoped that they'd move faster on that part but even the macos port is still not public yet, though at least MS Teams iirc is using it now internally at least. Afaik _planning_ for Linux still haven't begun.

> Seems like the latest webkitgtk shipped with gtk 4 support so there might be some improvements going there.

Yeah, we were told that it's better, but the difference seems to be tiny. 🤏 Also it's not a trivial upgrade path on its own already, and additionally requires a complete rewrite of our tray implementation (because libappindicator requires gtk3) among other things.

> But before that ships in the OS it is going to be years.

True, but with our switch to webkitgtk 4.1 we've raised the minimum distro version already to the same versions that ship the gtk4 version too.

> CEF work I think is the way to go. Even better if the user can patch the chromium version without us. I know that for security oriented OS (like Parrot OS) the maintenance of electron apps is annoying because they cant easily patch the browser in case there is a critical CVE.

Yeah, ideally we could create a webview library out of it because each cef update by itself is technically a breaking change and could always break apps (to our current knowledge, though i hope i'm wrong and we can just share the lib files themselves between apps 🙏 ). Due to the urgentness of the situation imo we should get _something_ working first even if the distribution method is not optimal. Appimages, our current main distribution format, require to bundle all cef libraries in it by design anyway.

**Comment by ahqsoftwares** [comment]
Fabian (not pinging cuz I'm quite a bit out of context with wry for a whole lotta time)
I guess wry should start ditching webkitgtk and get chromium before tauri v2 drops in, at least for linux. I wish fiefox had its own webview, its shipped in almost all desktop distros

**Comment by AlexProgrammerDE** [comment]
There is a open feature request on Firefoxes feature request tracker: https://connect.mozilla.org/t5/ideas/create-an-alternative-to-electron/idi-p/268
It hasn't received much attention by the Firefox team apparently though.

**Comment by FabianLars** [comment]
> Fabian (not pinging cuz I'm quite a bit out of context with wry for a whole lotta time)
> I guess wry should start ditching webkitgtk and get chromium before tauri v2 drops in, at least for linux. I wish fiefox had its own webview, its shipped in almost all desktop distros

We can't do it for / before v2. It's simply too much work and we don't want to hold back v2 any longer.

**Comment by FabianLars** [comment]
> There is a open feature request on Firefoxes feature request tracker: https://connect.mozilla.org/t5/ideas/create-an-alternative-to-electron/idi-p/268
> It hasn't received much attention by the Firefox team apparently though.

While a Firefox alternative would be insanely nice, I highly doubt that we'll ever see anything close to it. Firefox' codebase is basically incompatible with this idea (and it's not like they're already overworked with just Firefox as a use case alone)

**Comment by simonhyll** [comment]
> Fabian (not pinging cuz I'm quite a bit out of context with wry for a whole lotta time) I guess wry should start ditching webkitgtk and get chromium before tauri v2 drops in, at least for linux. I wish fiefox had its own webview, its shipped in almost all desktop distros

Just wanna underscore that CEF definitely won't be in v2, hopefully it'll be in Tauri v3 or v4, depending on how fast Microsoft decides to come through on their promise of shipping WebView2 for Mac and Linux.

Firefox definitely not happening since it's a browser not a webview, and Servo is years away from being an improvement over webkitgtk.

**Comment by ahqsoftwares** [comment]
> > Fabian (not pinging cuz I'm quite a bit out of context with wry for a whole lotta time) I guess wry should start ditching webkitgtk and get chromium before tauri v2 drops in, at least for linux. I wish fiefox had its own webview, its shipped in almost all desktop distros
> 
> Just wanna underscore that CEF definitely won't be in v2, hopefully it'll be in Tauri v3 or v4, depending on how fast Microsoft decides to come through on their promise of shipping WebView2 for Mac and Linux.
> 
> Firefox definitely not happening since it's a browser not a webview, and Servo is years away from being an improvement over webkitgtk.

I meant to just get the official cef that electron uses, at least on linux rn (ik there's a lot into it with the fact that most of cef releases is breaking)

**Comment by simonhyll** [comment]
> I meant to just get the official cef that electron uses, at least on linux rn (ik there's a lot into it with the fact that most of cef releases is breaking)

Electron doesn't use CEF, it's literally a custom build of the full Chromium browser. And there is no "just" with it, it's a huge undertaking, it can't be done for v2 both from an amount of work point of view and a security audit point of view. Earliest it can possibly happen is in Tauri version 3

**Comment by ahqsoftwares** [comment]
@simonhyll Seems like I'm quite a bit out of context 😅 about the current state of cef, electron and tauri

**Comment by westurner** [comment]
- [ ] WebView with CEF: Chromium Embedded Framework

- CEF: Chromium Embedded Framework
  - Wikipedia: https://en.wikipedia.org/wiki/Chromium_Embedded_Framework
  - Docs: https://bitbucket.org/chromiumembedded/cef/wiki/Home
  - Docs: https://bitbucket.org/chromiumembedded/cef/wiki/SandboxSetup.md
    > CEF supports the Chromium sandbox on each platform as described below.
  - CEF > External Projects:
    https://github.com/chromiumembedded/cef#external-projects
    - [x] cefpython: https://github.com/cztomczak/cefpython
    - [ ] https://github.com/wusyong/cef-rs
    - [ ] ChromiumGtk WebView.cs: https://github.com/lunixo/ChromiumGtk/blob/main/Lunixo.ChromiumGtk/WebView.cs
      - How much work would it be to create a WebView interface implementation compatible with WRY bsed on e.g. cef-rs?
        - webview2 is Edge on windows only, and Edge is Chromium-based:
          https://github.com/tauri-apps/wry/blob/dev/src/webview2/mod.rs

**Comment by Bricklou** [comment]
Would be there a way to use CEF on linux without embedding it? (By taking advantage of already installed browser on the system like chrome/chromium, ms edge, etc.)

**Comment by FabianLars** [comment]
CEF? No, cef is a seperate thing and not part of any browser.
Chromium in general? Kindaaa, but also not really. Browsers aren't webviews so we would be heavily limited and would need a second different IPC implementation.

**Comment by ahqsoftwares** [comment]
> CEF? No, cef is a seperate thing and not part of any browser. Chromium in general? Kindaaa, but also not really. Browsers aren't webviews so we would be heavily limited and would need a second different IPC implementation.


Yeah, cef has to be made as a library like webkitgtk for it to work system wide

**Comment by ahqsoftwares** [comment]
The deal breaker is that every release of cef can be breaking

**Comment by alexoceann** [comment]
Hi all, has any consideration been given to using Deno's `rusty_v8` bindings (or `deno_core`) as an optional renderer? It seems to be a good fit for the issue here.

- [rusty_v8](https://github.com/denoland/rusty_v8)
- [Announcing Stable V8 Bindings for Rust](https://deno.com/blog/rusty-v8-stabilized)
- [deno_core](https://github.com/denoland/deno_core)
- [roll-your-own-javascript-runtime](https://github.com/denoland/roll-your-own-javascript-runtime)

**Comment by FabianLars** [comment]
not really because we need more than just v8. v8 is not a renderer or something like a webview, it's just the js runtime, so one (arguably large) piece of the puzzle.

**Comment by mootari** [comment]
> depending on how fast Microsoft decides to come through on their promise of shipping WebView2 for Mac and Linux

On July 6 they announced [the following](https://github.com/MicrosoftEdge/WebView2Feedback/issues/1314#issuecomment-2211683486):
> After careful consideration and review of our long-term product strategy, we have decided to discontinue plans for a public release of WebView2 on MacOS and Linux to focus on providing the most value to our customers on currently supported platforms.

**Comment by FabianLars** [comment]
Yep, we're aware (and tbh we weren't that hopeful a while before that announcement anymore either). I thought we linked that announcement here but I guess I messed up so thanks for doing that!

With this we're officially back to primarily CEF. I don't have anything to share except that we started talking about it a few days ago again x) I hope we'll have at least some kind of starting point or approach to share after holidays but we'll see...

We're also keeping a close eye on https://github.com/chromiumembedded/cef/issues/3836 if someone's interested in that.

P.S. Servo of course is not forgotten but it is obviously not a proper replacement for many use cases still

**Comment by WesWeCan** [comment]
Completely different approach (maybe stupid approach), I worked some time ago with Puppeteer. And there was an option to find a local installation of chrome and run puppeteer from the already installed instance.

Can this not be an option? to hook in on a already existing browser.

**Comment by westurner** [comment]
browser_cookie3 has a bunch of paths for ChromiumBased browsers:
```
class ChromiumBased:
class Chrome(ChromiumBased):
class Arc(ChromiumBased):
class Chromium(ChromiumBased):
class Opera(ChromiumBased):
class OperaGX(ChromiumBased):
class Brave(ChromiumBased):
class Edge(ChromiumBased):
class Vivaldi(ChromiumBased):
```

https://github.com/borisbabic/browser_cookie3/blob/03895797e48dd107806db171d8392c562151807d/browser_cookie3/__init__.py#L431

**Comment by oatmealproblem** [comment]
> Completely different approach (maybe stupid approach), I worked some time ago with Puppeteer. And there was an option to find a local installation of chrome and run puppeteer from the already installed instance.
> 
> Can this not be an option? to hook in on a already existing browser.

According to Fabian, no: https://github.com/tauri-apps/tauri/issues/14963#issuecomment-3915822953

See also the Gluon project, which attempted that, but the creator determined that it would probably never work reliably

**Comment by mxiloyannis** [comment]
Huge fun or Tauri here,  we use it for almost all of our internal tools at Akina (where I work). When it came to decide on the technology to bundle our web app into a Desktop app, however, we had to go with Electron because we cannot afford testing and providing support for the different OSs. We're a small team with limited resources. 
I have been following this conversation and love where things are going. If there a was an option to include Servo or embedding chromium in Tauri with CEF, it would really be a no-brainer to choose Tauri. 
I'd be also happy to help if I can.

**Comment by adrian-afl** [comment]
Honestly if you need to bundle chromium with Tauri for it to be really cross platform compatible why not use Electron at this point?

**Comment by sleeyax** [comment]
> Honestly if you need to bundle chromium with Tauri for it to be really cross platform compatible why not use Electron at this point?

I mean, for literally all the other benefits Tauri offers over Electron.

**Comment by Norbiros** [comment]
The goal is to make CEF an opt-in alternative to WebKit, likely targeting Linux since other platforms have superior webviews compared to WebKit. Additionally, work is underway to make CEF a shared dependency: https://github.com/chromiumembedded/cef/issues/3836, so binary size still could be small...

**Comment by ahqsoftwares** [comment]
> The goal is to make CEF an opt-in alternative to WebKit, likely targeting Linux since other platforms have superior webviews compared to WebKit. Additionally, work is underway to make CEF a shared dependency: [chromiumembedded/cef#3836](https://github.com/chromiumembedded/cef/issues/3836), so binary size still could be small...

The main problem is that, every **release** of cef is breaking. And yeah, its kinda hard to integrate well

**Comment by FabianLars** [comment]
that's what the linked issue is trying to solve

**Comment by Ruben-Kruepper** [comment]
As a MacOS first shop I'd like to add that this isn't just a Linux issue - MacOS webviews are also wildly inconsistent.

Making the bundle just a bit smaller for a desktop app is outweighed by the simplicity and consistency of having Chromium bundled for many users imho.

The final straw is the WebKit devtools 😭

**Comment by ahqsoftwares** [comment]
I guess iOS also suffers from the same webview issue. Though on the apple side, you may use html, js, css features that are available on the lowest macos version of safari.

Seems like Windows 10, 11 only has auto updating webview2 sadly


(Please Forgive my broken English)

**Comment by zkvsky** [comment]
The solution is to to go with chromium until servo/verso gets stable enough, which will very likely happen earlier than webkitgtk becoming usable.

**Comment by Norbiros** [comment]
I agree. In future there will be more engines like LibWeb from the ladybird project etc, but for now chromium is the most stable and usable for everyone...

**Comment by ammelto** [comment]
I'm so excited for the day Tauri gets bundled with CEF. The value of Tauri for me has always been the rust backend, not the small bundle size. On Mac and Linux it's a nightmare trying to maintain feature parity if you're doing anything remotely complex. We're building a CAD tool and doing any kind of 3D rendering on a Mac or Linux build target is basically impossible. WebGL in CEF is extremely performant, and basically non-existent in Safari. We're trying everything we can not to switch to electron, but (as silly as it might sound) electron has better performance for just that rendering pipeline. Though, we really don't want to sacrifice our rust backend just for that since it'll cause all other sorts of problems for us.

Anyway, commenting for support :)

**Comment by westurner** [comment]
How is electron faster at WebGL than CEF? (Sorry OT)

**Comment by ammelto** [comment]
> How is electron faster at WebGL than CEF? (Sorry OT)
> […](#)

Other way around. Electron is faster than Tauri as it stands currently for doing any kind of 3D render through webgl. Tauri using CEF would fix that.

Dioxus is also showing some promise with the new [WGPU Overlays](https://dioxuslabs.com/blog/release-060#hybrid-wgpu-overlays).

**Comment by aervxa** [comment]
CEF in Tauri will be revolutionary!!

commenting for support  ;') cuz i don't have the knowledge to make a PR to fix the issue 😭 (it won't be that easy)

**Comment by MulverineX** [comment]
> > Fabian (not pinging cuz I'm quite a bit out of context with wry for a whole lotta time) I guess wry should start ditching webkitgtk and get chromium before tauri v2 drops in, at least for linux. I wish fiefox had its own webview, its shipped in almost all desktop distros
> 
> Just wanna underscore that CEF definitely won't be in v2, hopefully it'll be in Tauri v3 or v4, depending on how fast Microsoft decides to come through on their promise of shipping WebView2 for Mac and Linux.
> 
> Firefox definitely not happening since it's a browser not a webview, and Servo is years away from being an improvement over webkitgtk.

@simonhyll 

webview2 on linux/mac is cancelled https://github.com/MicrosoftEdge/WebView2Feedback/issues/645#issuecomment-2211683674

**Comment by TCROC** [comment]
We noticed serious performance issues on our Linux builds.  I'm glad we discovered this in the first few days of starting our project rather than getting farther down the road.  I'm just checking in now to see what the status is on CEF / Servo / alternate webview / runtime?  I'd love to be able to stick with tauri for Linux if we can.

**Comment by FabianLars** [comment]
> what the status is on CEF

https://github.com/tauri-apps/tauri/tree/feat/cef - currently more like source available than open source since most of the development happens at work.

> what the status is on Servo

https://v2.tauri.app/blog/tauri-verso-integration/ - we're trying to keep the current state working but further development is halted for now since we're missing funding and/or manpower for that

> what the status is on alternative webview

we're also back looking at qtwebengine as a replacement for webkitgtk, work is happening here https://github.com/wusyong/cxx-qt-widgets but it'll be paused for a while since Yu-Wei has to focus on his work for a while.

---

We're currently also thinking to keep Wry a system webview library (after all Wry stands for WebView Rendering Library) so i will move this issue to the tauri repository instead.

**Comment by TCROC** [comment]
@FabianLars thank you for the quick response!  Are there any quick instructions to test out the cef branch?  I'd love to be an early tester for you if its ready :)

**Comment by FabianLars** [comment]
Some of the repo examples should work. It should mostly work without much setup, just point your tauri deps in Cargo.toml to the branch (i recommend a patch.crates-io section so it's replaced in plugins as well), enable the "cef" feature flag, and use `cargo install tauri-cli --git https://github.com/tauri-apps/tauri --branch feat/cef`

But yeah, other than that you're a bit on your own atm, but hopefully not for that much longer :)

**Comment by siriux** [comment]
> we're also back looking at qtwebengine as a replacement for webkitgtk, work is happening here https://github.com/wusyong/cxx-qt-widgets but it'll be paused for a while since Yu-Wei has to focus on his work for a while.

This could be a lightweight alternative on Linux, as for many users it could be considered a system WebView already provided by their desktop.

And, in this case, it would make sense to integrate it into Wry, as the QT libraries could be provided directly by the linux distribution, as opposed to CEF where it normally would be embedded into the application.

**Comment by FabianLars** [comment]
> This could be a lightweight alternative on Linux, as for many users it could be considered a system WebView already provided by their desktop.

Exactly :)

> And, in this case, it would make sense to integrate it into Wry, as the QT libraries could be provided directly by the linux distribution, as opposed to CEF where it normally would be embedded into the application.

Yeah, i had similar thoughts but it doesn't look like it'll be straight-forward so we cannot make promises that this will happen.
Edit: webkitgtk tightly coupled to gtk + qtwebengine tightly coupled to qt = likely complete rewrite of the linux implementation of both wry and tao, if that even works with how different qt works (as-in whether can be integrated into wry).

**Comment by the-watchmaker** [comment]
Looking forwards to how it's gonna pan out. In our project, we just added an Electron sidecar to make some of the features work for the time being. Pretty seamless but uncomfortable. It's like attending your cousins wedding, wearing regular pants on top of wet swimming pants right after the pool party.

**Comment by brycedbjork** [comment]
this would take tauri to the next level: real devtools, more consistent rendering, better embedded browsers. so excited for this to ship!!

**Comment by yuyang-ok** [comment]
any progress on this issue?

**Comment by cfc23** [comment]
Any word on the status of CEF? I see there's a branch, but how close to usable is it?

**Comment by FabianLars** [comment]
No news. Ya'll don't have to keep asking. If there are news I will share them. And yes, the branch is mostly usable but like I said above, you're on your own when you want to try it. We cannot yet offer much help (please do report issues though if you do try it).

**Comment by Pinguin2001** [comment]
A key point to consider is that many devs adopt Tauri specifically for its tiny bundle size. Shipping Chromium, while solving a few issues, inevitably increases storage consumption and make the experience worse for everybody.

I don't think simply treating user storage as 'free real estate' by shipping large applications (like 200+MB) is the best long-term solution. While I see the issues with WebKitGTK, I hope this remains Linux-exclusive. This makes sense because Tauri already works excellently on macOS and Windows using the native renderers WKWebView and WebView2.

Taking readest as an example, you would seriously add CEF to the app because the required feature in WKWebView is not available in macOS 12? This version is used by a few percent of all Mac users and is five years old.

**Comment by ds82** [comment]
> I don't think simply treating user storage as 'free real estate' by shipping large applications (like 200+MB) is the best long-term solution. 

Let the developer decide which tradeoff he wants to make .. exactly the approach https://zero-native.dev/ uses.

**Comment by Bricklou** [comment]
If the chromium bundle is installed once (by sharing it with others apps) and is not bundled inside the app binary, I don't see the issues of working with CEF.

**Comment by FabianLars** [comment]
And that's the thing here. In the foreseeable future the CEF instances are **not** shared across apps. This is something we really really want of course but it's not exactly a simple task. Main upstream issue is this one i think https://github.com/chromiumembedded/cef/issues/3836

**Comment by Revertron** [comment]
> Tauri already works excellently on macOS and Windows

I wouldn't call it excellently when a small app runs 6 processes totalling in 400MB of RAM.

But other your points are very valid!

**Comment by justbispo** [comment]
> And that's the thing here. In the foreseeable future the CEF instances are **not** shared across apps. This is something we really really want of course but it's not exactly a simple task. Main upstream issue is this one i think [chromiumembedded/cef#3836](https://github.com/chromiumembedded/cef/issues/3836)

Recently Arch Linux added a [CEF](https://archlinux.org/packages/extra/x86_64/cef/) package to their repos. Currently it's only used by the [OBS Studio browser plugin](https://archlinux.org/packages/extra/x86_64/obs-studio-plugin-browser/). Do you know if this is effectively a build of CEF that can be shared by multiple apps, or something very specific to the OBS plugin?

**Comment by FabianLars** [comment]
it's a step in the right direction i'd say. ignoring that it's just arch, the problem is that it does not solve the api versioning issue (Phase 2 in https://github.com/chromiumembedded/cef/issues/3836). Which seems fine for their usecase (though i see no mechanism that makes sure it doesn't break on updates) but sucks for Tauri as it would require re-compiling the app for the specific cef version that's in the repo.

**Comment by MavenRain** [comment]
For folks watching this thread who are interested in a focused-subset pure-Rust runtime as a third option alongside Chromium-via-CEF and Servo-via-Verso, I posted a Show-and-tell for `tauri-runtime-servocat` [here](https://github.com/orgs/tauri-apps/discussions/15546).  It is not a Chromium or Servo replacement, but it aims at the niche where neither is the right fit (settings panels, splash screens, embedded UIs).  Feedback is welcome.

**Comment by LaneSun** [comment]
In recent months, I've seen quite a few apps switch from Tauri to Electron — OpenCode, for example. Although they didn't go into detail about the reasons, based on my experience using the Tauri version of OpenCode on Linux (the scrolling experience was extremely poor, the top bar was ugly, updates didn't work properly, some sound effects wouldn't play — everything got better after switching to Electron), they were undoubtedly disappointed with Tauri. It's not that there were no particular reasons; rather, there were too many reasons to list.

This isn't an isolated case, and I think one of the reasons for this phenomenon is this very Issue.

**Comment by LaneSun** [comment]
To be honest, I don't like Electron at all. Its API design isn't great (I've accepted the downside of large bundle sizes, but the API keeps torturing me). So when I first tried Tauri 3–4 years ago, I really liked it — the development experience was unparalleled (except for Rust's huge build footprint).

But then I started running into various roadblocks: it wouldn't launch on Linux, and when it did, certain features and behaviors were inconsistent. All sorts of issues, all centered around Linux. For a while, I used Windows for development, and it worked great — no problems at all. (I don't use Mac; Linux is my daily system.)

You can imagine how frustrated I was. Basically, Tauri was nearly unusable for me (you wouldn't accept a framework that occasionally just shows a blank page due to system WebKit issues). After that, I bounced back and forth between Tauri and Electron, then I found this issue and started following it.

And now years have passed! I still see no change. Tauri's development experience on Linux remains terrible. I might be getting emotional — I'm just complaining, nothing more. After all, this might be the last time I check this issue. I just want the developers to know that people like me once existed.

**Comment by chrox** [comment]
> And now years have passed! I still see no change. Tauri's development experience on Linux remains terrible. I might be getting emotional — I'm just complaining, nothing more. After all, this might be the last time I check this issue. I just want the developers to know that people like me once existed.

We share the same frustration in [Readest](https://github.com/readest/readest). Linux WebKitGTK issues have been one of our biggest pain points, causing almost all issues on Linux.

We’re planning to try the Tauri CEF branch soon and will report back with our findings: https://github.com/tauri-apps/tauri/tree/feat/cef

**Comment by LaneSun** [comment]
Deno released version 2.9, which adds a feature called [`deno desktop`](https://deno.com/blog/v2.9#deno-desktop) that directly converts a server script into a desktop application. Although it does not have native Rust support (it can be supported through some means, but I recall there is no native Rust support), its interface and implementation are the most elegant I have seen so far. It also supports dual Webview/CEF modes. This feature should currently be in beta (based on my understanding of the Deno team's usual practices, even beta versions are generally quite usable), so I am only mentioning it here in case some people urgently need a third alternative for JS/TS desktop application platforms.

**Comment by cmtheit** [comment]
I ship a tauri-runtime implementation backed by CEF in [kabegame](https://github.com/kabegame/kabegame) (released, desktop standard build on Windows/macOS/Linux) — sharing it as a real-world data point for this thread, not as a general-purpose solution.

Honest scope: it's windowed CEF Views (Chromium owns the top-level window + BrowserView), and it pins a CEF version, so — as noted above re: the CEF API-versioning problem — the app has to be recompiled per CEF version. That's fine for a first-party app, not a fix for Tauri's general QA/versioning concern.

Where it might actually help others going down this road is the pile of platform gotchas we had to solve, written up here: [tauri-runtime-cef/README](https://github.com/kabegame/kabegame/blob/main/src-tauri/tauri-runtime-cef/README.md) — e.g. macOS MachPortRendezvous bundle-id consistency (window shell fine but content black otherwise), the flat-subprocess-path CEF patch vs. the 5 helper .app variants, the Windows app-manifest requirement for the GPU process's layered child window (CEF #3765), and per-platform custom-scheme handling (http://tauri.localhost rewriting → ERR_CONNECTION_REFUSED if you register the raw custom scheme). Happy to answer questions.

**Comment by MulverineX** [comment]
I think you linked the wrong PR in CEF @cmtheit

**Comment by MulverineX** [comment]
Translated the kabegame tauri-runtime-cef README in chunks using DeepL: [gist.github.com/MulverineX@kabegame-tauri-runtime-cef.md](https://gist.github.com/MulverineX/4a26a5d6b20b91856c665655be7ce1eb)

**Comment by bra1n** [comment]
> Happy to answer questions.

@cmtheit it would be great if you could write a brief guide on how to get from a stock Tauri "Hello World" app to one that uses CEF. Your gotcha write-up is great, but also really confusing for someone who is not yet using CEF in their own project.

**Comment by westurner** [comment]
> WebkitGTK

WebkitGTK spawns with bwrap unless already in a flatpak;

> It looks like WebKitGTK uses bubblewrap on linux unless running in a
flatpak, because flatpaks don't have permission to create namespaces so
bwrap can't run.

> - WebKit/Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp at main · WebKit/WebKit https://github.com/WebKit/webkit/blob/main/Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp

> - WebKit/Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.cpp at main · WebKit/WebKit https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.cpp

> dloss/awesome-agent-sandboxes: A curated list of sandboxing solutions for
AI agents https://github.com/dloss/awesome-agent-sandboxes

**Comment by westurner** [comment]
It looks like you haven't understood the relavance of sandboxing when packaging CEF?

~~Not sure why those links are ellipsized. EDIT: thanks gmail to github~~

If you package CEF in a flatpak, will bwrap fail there too?

**Comment by The-Best-Codes** [comment]
I migrated to the feat/cef branch on my Tauri app temporarily, but wound up migrating back since it wasn't a smooth experience. It's certainly coming along though! Here's my commit in case it helps anyone:
https://github.com/The-Best-Codes/agent-one/commit/30016f027047e28604ef7624c84d6c8d01826afe

**Comment by FabianLars** [comment]
@The-Best-Codes Can you tell us about the issues you had please?

**Comment by The-Best-Codes** [comment]
@FabianLars I thought that last time I had run `bun tauri dev` and needed to install some extra prerequisites, but it worked fine this time. It does crash when I try to open the inspector (via Ctrl + Shift + I shortcut, as the context menu appears to be disabled), but there isn't any error message in the terminal.
I'll have to explore more later but that's all I remember off the top of my head!
